platform/linux-generic/odp_pcapng.c

```diff
@@ -0,0 +1,434 @@
+			       odp_global_data.inotify_watch_fd);
+	if (ret)
+		ODP_ERR("can't deregister inotify %s\n", strerror(errno));
+
+	if (odp_global_data.inotify_pcapng_fd != -1)
+		close(odp_global_data.inotify_pcapng_fd);
+
+	if (odp_global_data.inotify_watch_fd != -1)
+		close(odp_global_data.inotify_watch_fd);
+
+	for (i = 0; i < max_queue; i++) {
+		char pcapng_name[128];
+		char pcapng_path[256];
+
+		entry->s.pcapng.state[i] = PCAPNG_WR_STOP;
+		close(entry->s.pcapng.fd[i]);
+
+		get_pcapng_fifo_name(pcapng_name, sizeof(pcapng_name),
+				     entry->s.name, i);
+		snprintf(pcapng_path, sizeof(pcapng_path), "%s/%s",
+			 PCAPNG_WATCH_DIR, pcapng_name);
+
+		if (remove(pcapng_path))
+			ODP_ERR("can't delete fifo %s\n", pcapng_path);
+	}
+}
+
+int write_pcapng_hdr(pktio_entry_t *entry, int qidx)
+{
+	size_t len;
+	pcapng_section_hdr_block_t shb;
+	pcapng_interface_description_block_t idb;
+	int fd = entry->s.pcapng.fd[qidx];
+
+	memset(&shb, 0, sizeof(shb));
+	memset(&idb, 0, sizeof(idb));
+
+	shb.block_type = PCAPNG_BLOCK_TYPE_SHB;
+	shb.block_total_length = sizeof(shb);
+	shb.block_total_length2 = sizeof(shb);
+	shb.magic = PCAPNG_ENDIAN_MAGIC;
+	shb.version_major = 0x1;
+	shb.version_minor = 0x0;
+	shb.section_len = -1;
+
+	len = write(fd, &shb, sizeof(shb));
+	/* fail to write shb/idb means the pcapng is unreadable */
+	if (len != sizeof(shb)) {
+		ODP_ERR("Failed to write pcapng section hdr\n");
+		return -1;
+	}
+	fsync(fd);
+
+	idb.block_type = PCAPNG_BLOCK_TYPE_IDB;
+	idb.block_total_length = sizeof(idb);
+	idb.block_total_length2 = sizeof(idb);
+	idb.linktype = PCAPNG_LINKTYPE_ETHERNET;
+	idb.snaplen = 0x0; /* unlimited */
+	len = write(fd, &idb, sizeof(idb));
+	if (len != sizeof(idb)) {
+		ODP_ERR("Failed to write pcapng interface description\n");
+		return -1;
+	}
+	fsync(fd);
+
+	return 0;
+}
+
+/*
+ * make sure that each fifo write is less than PIPE_BUF
+ * this will make sure writes are atomic (on non blocking mode).
+ * writev() transfers all the data and returns the number of bytes requested or
+ * -EAGAIN
+ */
+static ssize_t write_fifo(int fd, struct iovec *iov, int iovcnt)
+{
+	ssize_t len = 0;
+
+	len = writev(fd, iov, iovcnt);
+	/*
+	 * we don't care if a writev fails, we asynchronously read the fifo
+	 * so the next block of packets might be successful. This error only
+	 * means that some packets failed to append on the pcap file
+	 */
+	if (len > 0)
+		fsync(fd);
+
+	return len;
+}
+
+int write_pcapng_pkts(pktio_entry_t *entry, int qidx,
+		      const odp_packet_t packets[], int num)
+{
+	int i = 0;
+	struct iovec packet_iov[3 * num];
+	pcapng_enhanced_packet_block_t epb[num];
+	int iovcnt = 0;
+	ssize_t block_len = 0;
+	int fd = entry->s.pcapng.fd[qidx];
+	ssize_t len = 0, wlen;
+
+	for (i = 0; i < num; i++) {
+		odp_packet_hdr_t *pkt_hdr = packet_hdr(packets[i]);
+		uint32_t seg_len;
+		char *buf = (char *)odp_packet_offset(packets[i], 0, &seg_len,
+						      NULL);
```

Comment:
Is writing only first segment enough?

Reply:
No but it won't break the capture. This was suggested by @Bill-Fischofer-Linaro as an improvement over the initial commit